src/nanoarrow/nanoarrow_device_cuda.c

```diff
@@ -392,6 +417,251 @@ static ArrowErrorCode ArrowDeviceCudaArrayMove(struct ArrowDevice* device_src,
+
+    // Don't initialize a buffer if we have already started the copy.
+    // This check works because the buffer size is set synchronously
+    // when the copy operation is launched.
+    if (buffer_dst->size_bytes > 0) {
+      continue;
+    }
+
+    // Otherwise, initialize the buffer
+    NANOARROW_RETURN_NOT_OK(ArrowDeviceCudaBufferInitAsync(
+        device_src, src_view, device_dst, buffer_dst, hstream));
+  }
+
+  // Recurse for children
+  for (int64_t i = 0; i < src->n_children; i++) {
+    NANOARROW_RETURN_NOT_OK(ArrowDeviceCudaArrayViewCopyBuffers(
+        device_src, src->children[i], device_dst, dst->children[i], hstream,
+        unknown_buffer_size_count));
+  }
+
+  // ...and for dictionary
+  if (src->dictionary != NULL) {
+    NANOARROW_RETURN_NOT_OK(ArrowDeviceCudaArrayViewCopyBuffers(
+        device_src, src->dictionary, device_dst, dst->dictionary, hstream,
+        unknown_buffer_size_count));
+  }
+
+  return NANOARROW_OK;
+}
+
+static ArrowErrorCode ArrowDeviceCudaCreateStream(struct ArrowDevice* src,
+                                                  struct ArrowDevice* dst, CUstream* out,
+                                                  struct ArrowError* error) {
+  struct ArrowDeviceCudaPrivate* private_data;
+
+  if (dst->device_type == ARROW_DEVICE_CUDA) {
+    private_data = (struct ArrowDeviceCudaPrivate*)dst->private_data;
+  } else if (src->device_type == ARROW_DEVICE_CUDA) {
+    private_data = (struct ArrowDeviceCudaPrivate*)src->private_data;
+  } else {
+    return NANOARROW_OK;
+  }
+
+  NANOARROW_CUDA_RETURN_NOT_OK(cuCtxPushCurrent(private_data->cu_context),
+                               "cuCtxPushCurrent", NULL);
+  CUresult err = cuStreamCreate(out, CU_STREAM_NON_BLOCKING);
+  CUcontext unused;
+  NANOARROW_CUDA_ASSERT_OK(cuCtxPopCurrent(&unused));
+  NANOARROW_CUDA_RETURN_NOT_OK(err, "cuStreamCreate", error);
+
+  return NANOARROW_OK;
+}
+
+static void ArrowDeviceCudaDestroyStream(CUstream hstream) {
+  if (hstream != NANOARROW_CUDA_DEFAULT_STREAM) {
+    NANOARROW_CUDA_ASSERT_OK(cuStreamDestroy(hstream));
+  }
+}
+
+static ArrowErrorCode ArrowDeviceCudaArrayViewCopyInternal(
+    struct ArrowDevice* device_src, struct ArrowDeviceArrayView* src,
+    struct ArrowDevice* device_dst, struct ArrowArray* dst, struct ArrowError* error) {
+  CUresult err;
+  int code;
+  int64_t unknown_buffer_size_count = 0;
+
+  // Create a stream to do the first (and maybe only) round of buffer copying.
+  // The ArrowDeviceArrayView should probably have a sync_event so that we can
+  // force the stream we just created to wait on the event.
+  CUstream hstream = NANOARROW_CUDA_DEFAULT_STREAM;
+  NANOARROW_RETURN_NOT_OK(
+      ArrowDeviceCudaCreateStream(device_src, device_dst, &hstream, error));
+
+  code = ArrowDeviceCudaArrayViewCopyBuffers(device_src, &src->array_view, device_dst,
+                                             dst, hstream, &unknown_buffer_size_count);
```

Comment:
If `src` has a sync event then we need to wait on that event in `hstream` otherwise we have a possible race here

Reply:
In hindsight, I think this is an issue across most of the functions in this PR where we always need to be checking the sync event in source arrays and waiting on them in the stream used for the function